File: imagine/events/regime_event_trigger.py
from imagine.events.imagine_event import ImagineEvent


class RegimeEventTrigger:

    def __init__(self, event_name, trigger, redefinable):

        self._private_trigger = trigger
        self.event_name = event_name
        self.regime_redefinable = redefinable
        self._private_redefined_trigger = None
        self._private_regime_redefined = False

    @property
    def trigger(self):
        if self.regime_redefined:
            return self._private_redefined_trigger
        else:
            return self._private_trigger

    @property
    def regime_redefined(self):
        return self._private_regime_redefined

    @regime_redefined.setter
    def regime_redefined(self, redefined_flag):
        if self.regime_redefinable:
            self._private_regime_redefined = redefined_flag

    @property
    def name(self):
        return self.event_name

    @trigger.setter
    def trigger(self, t):
        if self.regime_redefined:
            self._private_redefined_trigger = t
        else:
            self._private_trigger = t

    def set_private_trigger(self, private_trigger):
        self._private_trigger = private_trigger

    def set_private_redefined_trigger(self, private_redefined_trigger):
        self._private_redefined_trigger = private_redefined_trigger

    def convert_to_event(self, status):
        ev = ImagineEvent(self.name, status)
        if status.regime_redefinable:
            ev.status.regime_redefined = False
            ev.trigger = self._private_trigger
            ev.status.regime_redefined = True
            ev.trigger = self._private_redefined_trigger
            ev.status.regime_redefined = self.regime_redefined
        else:
            ev.trigger = self.trigger
        return ev

    def convert_from_event(self, ev):
        self.event_name = ev.name
        self.regime_redefinable = ev.status.regime_redefinable
        if self.regime_redefinable:
            self.regime_redefined = ev.status.regime_redefined
            ev.status.regime_redefined = False
            self._private_trigger = ev.trigger
            ev.status.regime_redefined = True
            self._private_redefined_trigger = ev.trigger
            ev.status.regime_redefined = self.regime_redefined
        else:
            self._private_trigger = ev.trigger

    def crop_name_has_changed(self, previous_name, new_name):
        if self._private_trigger is not None:
            self._private_trigger.crop_name_has_changed(previous_name, new_name)
        if self._private_redefined_trigger is not None:
            self._private_redefined_trigger.crop_name_has_changed(previous_name, new_name)

File: Matlab Code/Classes/RegimeEventTrigger.m
classdef RegimeEventTrigger
    %RegimeEventTrigger Provides a standard data structure to manage
    %triggers defined through the regime.
    %
    %   Detailed explanation goes here

    properties (Dependent)
       trigger 
       regimeRedefined
       name
    end
    
    properties
        eventName                
    end
    
    properties (Access = private)
        privateTrigger
        privateRedefinedTrigger
        regimeRedefinable        
        privateRegimeRedefined
    end
    
    methods
       
        function obj = RegimeEventTrigger(eventName, trigger, redefinable)
           if nargin < 3 
               error('RegimeEventTrigger constructor requires 3 arguments.');
           end
           obj.privateTrigger = trigger;
           obj.eventName = eventName;
           obj.regimeRedefinable = redefinable;
           obj.privateRegimeRedefined = false;
         end
        
        % Get one of the private triggers depending on the regimeRedefined
        % flag.
        function t = get.trigger(obj)
            if obj.regimeRedefined
                t = obj.privateRedefinedTrigger;
            else
                t = obj.privateTrigger;
            end
        end
        
        function obj = setPrivateTrigger(obj, privateTrigger)
           obj.privateTrigger = privateTrigger; 
        end
        
        function obj = setPrivateRedefinedTrigger(obj, privateRedefinedTrigger)
            obj.privateRedefinedTrigger = privateRedefinedTrigger;
        end
        
        function obj = set.regimeRedefined(obj, redefinedFlag)
            if obj.regimeRedefinable
               obj.privateRegimeRedefined = redefinedFlag; 
            end
        end
        
        function redefinedFlag = get.regimeRedefined(obj)
            redefinedFlag = obj.privateRegimeRedefined;
        end
        
        function name = get.name(obj)
            name = obj.eventName;
        end
        
        function obj = set.trigger(obj, t)
           if obj.regimeRedefined
              obj.privateRedefinedTrigger = t; 
           else
              obj.privateTrigger = t;               
           end
        end
                
        % status should come directly from the crop.
        % In the regime, we need to maintain, and overwrite the
        % regimeRedefined flag.
        function ev = convertToEvent(obj, status)
           
            % Want to set both private triggers.
            ev = ImagineEvent(obj.name, status);
            
            if status.regimeRedefinable
                ev.status.regimeRedefined = false;
                ev.trigger = obj.privateTrigger;
                ev.status.regimeRedefined = true;
                ev.trigger = obj.privateRedefinedTrigger;
                ev.status.regimeRedefined = obj.regimeRedefined;
            else
                ev.trigger = obj.trigger;
            end        
        end
       
        % Absorbs the event's data into this object.
        function obj = convertFromEvent(obj, ev)

            obj.eventName = ev.name;
            
            obj.regimeRedefinable = ev.status.regimeRedefinable;
            if obj.regimeRedefinable
                obj.regimeRedefined = ev.status.regimeRedefined;
                ev.status.regimeRedefined = false;
                obj.privateTrigger = ev.trigger;
                ev.status.regimeRedefined = true;
                obj.privateRedefinedTrigger = ev.trigger;   
                ev.status.regimeRedefined = obj.regimeRedefined; % Should not be necessary, but seems polite.
            else
                obj.privateTrigger = ev.trigger;
            end
            
        end
        
        function cropNameHasChanged(obj, previousName, newName)
            if ~isempty(obj.privateTrigger)
               obj.privateTrigger.cropNameHasChanged(previousName, newName);               
            end
            if ~isempty(obj.privateRedefinedTrigger)
               obj.privateRedefinedTrigger.cropNameHasChanged(previousName, newName);               
            end                
        end
    end
    
end

